### DISCUSSION TITLE
Preludes?

### DISCUSSION THREAD

**Discussion by killercup** [rust-lang-nursery/api-guidelines#140] [API Guidelines]
In general, it seems that the usage of glob imports (`use foo::*;`) is generally discouraged in Rust [[1]], while at the same time the std lib and some other crates offer "prelude" modules [[2]] that users are supposed to glob import for convenience.

Should there be a guideline describing when/whether a prelude is good to have?

[1]: https://doc.rust-lang.org/book/first-edition/crates-and-modules.html
[2]: https://github.com/search?utf8=✓&q=mod+prelude+language%3ARust&type=Code

- - -

Sorry if this is not something that is considered "API" enough for the API guidelines.

Context: https://github.com/killercup/assert_cli/pull/75#issuecomment-340608533 made me look up what the API guidelines say, and I couldn't find anything on the topic.

**Comment by emk** [comment]
There are definitely some APIs—especially for things like testing libraries or `diesel`—where I'm going to have to import the same 6 symbols every time I use the library. Usually this is a mix of traits and a few essential functions or types.

Personally, I'm rather fond of glob imports for cases like these, because otherwise I have to spend a lot more time digging around in the docs and figuring out errors just to get things done. Yes, I understand that in some sense it's not "ideal", because a new export might someday be added to the glob and conflict with some other import. But that's absolutely a risk I'm willing to pay in the name of convenience for certain carefully-chosen APIs.

I _do_ have a rule of "no glob imports for things which _aren't_ preludes", though. That seems pretty sensible.

**Comment by peter-lyons-kehl** [comment]
Misc. related notes:
- https://www.reddit.com/r/rust/comments/aw69h7/use_xy_as_can_be_used_as_prelude/ for re-exporting **multiple** items as `_`
  - good for extension traits.
  - especially good if you have (sets of) multiple extension traits with same names, for example
    - each repetition of the trait name is per `cfg` and/or feature setting/subset (so as to avoid pollution from conditional compilation clauses all over the code, and having conditional compilation once per extension submodule only). See `DebugExt` in https://github.com/peter-lyons-kehl/dis/blob/main/src/ext/core.rs and in https://github.com/peter-lyons-kehl/dis/blob/main/src/ext/proc_macro2_diagnostics.rs.
    - if your extension traits are generated by macro or by some other automation.
- `clippy` lint `wildcard_imports` has special handling: "allowed from modules that their name contains prelude." https://rust-lang.github.io/rust-clippy/master/#wildcard_imports
  - (imports that are public re-exports like `pub use some::path::*` don't trigger this lint by default)
  - if `clippy.toml` contains `warn-on-all-wildcard-imports = true`, then there's no special treatment of public re-exports OR paths that include "prelude" - then they trip this lint
  - `wildcard_imports` lints an import only if you actually use at least one item from the wildcard.
  ```rust
  #![deny(clippy::wildcard_imports)]
  
  mod m  {
    pub const C: () = {};
  }
  mod ext_prelude{
    pub mod core {
        pub const CC: () = {};
    }
  }
  
  use m::*;
  use ext_prelude::core::*; // substring "prelude" anywhere on the path, not necessarily the leaf module
  
  pub const USED_C: () = C; // this is needed to trip the lint
  
  pub const USED_CC: () = CC; // this works
  ```
